buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/rerun/executer/RerunTestExecuter.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal.test.rerun.executer;
+
+import org.elasticsearch.gradle.internal.test.rerun.TestRerunTaskExtension;
+import org.gradle.api.GradleException;
+import org.gradle.api.internal.tasks.testing.JvmTestExecutionSpec;
+import org.gradle.api.internal.tasks.testing.TestDescriptorInternal;
+import org.gradle.api.internal.tasks.testing.TestExecuter;
+import org.gradle.api.internal.tasks.testing.TestResultProcessor;
+import org.gradle.internal.id.CompositeIdGenerator;
+import org.gradle.process.internal.ExecException;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public final class RerunTestExecuter implements TestExecuter<JvmTestExecutionSpec> {
+
+    private final TestRerunTaskExtension extension;
+    private final TestExecuter<JvmTestExecutionSpec> delegate;
+    private TestJvmCrashReporter jvmCrashReporter;
+
+    public RerunTestExecuter(TestRerunTaskExtension extension, TestExecuter<JvmTestExecutionSpec> delegate) {
+        this.extension = extension;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void execute(JvmTestExecutionSpec spec, TestResultProcessor testResultProcessor) {
+        int maxRetries = extension.getMaxRetries().get();
+        if (maxRetries <= 0) {
+            delegate.execute(spec, testResultProcessor);
+            return;
+        }
+
+        RerunTestResultProcessor retryTestResultProcessor = new RerunTestResultProcessor(testResultProcessor);
+
+        int retryCount = 0;
+        JvmTestExecutionSpec testExecutionSpec = spec;
+
+        while (true) {
+            try {
+                delegate.execute(testExecutionSpec, retryTestResultProcessor);
+                break;
+            } catch (ExecException e) {
+                if (retryCount == maxRetries) {
+                    throw new GradleException("Max retries hit", e);
```

Comment:
So if we hit the retry limit we never call `storeActiveDescriptors()`. I think that would mean the crash reporter would report the wrong failures, should there have been any in the last round.

Reply:
Also, what's the benefit of the indirection of having a finalizing task action report the failures rather than just doing so right here? Is this code executed in the worker JVM vs the daemon or something?

Reply:
changed this to report the trace for each run and simplified things here.